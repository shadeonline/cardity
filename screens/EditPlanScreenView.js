import React, { useState, useEffect } from 'react';
import { useNavigation } from '@react-navigation/core';
import {
    StyleSheet,
    Text,
    View,
    TextInput,
    TouchableOpacity,
    Alert,
    KeyboardAvoidingView,
} from 'react-native';
import { firebaseEditLoyaltyPlan } from '../firebaseFunctions.js';
import { Ionicons } from '@expo/vector-icons';
import ColorSelector from "../components/ColorSelector.js";

const EditPlanScreenView = ({ route }) => {
    const navigation = useNavigation();



    const { programData } = route.params;
    const [editedDescription, setEditedDescription] = useState(programData.description);
    const [editedCardName, setEditedCardName] = useState(programData.loyaltyCard.cardName);
    const [editedStoreName, setEditedStoreName] = useState(programData.storeName);
    const [editedColor, setEditedColor] = useState(programData.loyaltyCard.color);
    const [editedTask, setEditedTask] = useState(programData.task);

    const formattedRewards = Object.keys(programData.rewards).map((stamps) => ({
        stamps,
        reward: programData.rewards[stamps],
    }));
    const [editedRewards, setEditedRewards] = useState(formattedRewards);

    const addRewardField = () => {
        setEditedRewards([...editedRewards, { stamps: '', reward: '' }]);
    };

    const removeRewardField = (index) => {
        const updatedRewards = [...editedRewards];
        updatedRewards.splice(index, 1);
        setEditedRewards(updatedRewards);
    };

    const handleEditPlan = async () => {
        if (
            !editedCardName ||
            !editedStoreName ||
            !editedDescription ||
            !editedRewards.some((reward) => {
                const stamps = parseInt(reward.stamps);
                return Number.isInteger(stamps) && stamps > 0 && reward.reward;
            })
        ) {
            Alert.alert('Validation Error', 'Please fill in all required fields correctly.');
            return;
        }
        const success = await firebaseEditLoyaltyPlan(
            programData.id,
            editedDescription,
            editedCardName,
            editedStoreName,
            editedColor,
            parseRewardsInput(editedRewards),
            editedTask
        );
        if (success) {
            // Plan created successfully, navigate to a success screen or handle as needed.
            console.log('Plan Edited!');
            navigation.goBack();
        } else {
            // Error occurred while creating the plan
            console.log('Plan not Edited!');
        }
    };

    const parseRewardsInput = () => {
        const parsedRewards = {};
        editedRewards.forEach((reward) => {
            if (reward.stamps && reward.reward) {
                parsedRewards[reward.stamps] = reward.reward;
            }
        });
        return parsedRewards;
    };


    return (
        <KeyboardAvoidingView style={styles.container} behavior="padding">

            <Text style={styles.tag}>Card Name:</Text>
            <TextInput
                style={styles.input}
                placeholder="i.e: Stag Nite Bar Card"
                value={editedCardName}
                onChangeText={(text) => setEditedCardName(text)}
            />

            <Text style={styles.tag}>Store Name:</Text>
            <TextInput
                style={styles.input}
                placeholder="i.e: Stag Nite Bar"
                value={editedStoreName}
                onChangeText={(text) => setEditedStoreName(text)}
            />

            <Text style={styles.tag}>Loyalty Plan Description:</Text>
            <TextInput
                style={styles.input}
                placeholder="i.e: Earn great and awesome reward with this plan"
                value={editedDescription}
                onChangeText={(text) => setEditedDescription(text)}
            />

            <Text style={styles.tag}>Task:</Text>
            <TextInput
                style={styles.input}
                placeholder="i.e: Earn great and awesome reward with this plan"
                value={editedTask}
                onChangeText={(text) => setEditedTask(text)}
            />

            <Text style={styles.tag}>Color of loyalty card:</Text>
            <View style={styles.colorPicker}>
                <ColorSelector onColorChange={setEditedColor} selectedColor={editedColor} />
            </View>

            {/* Input fields for rewards */}
            <Text style={styles.tag}>Rewards:</Text>
            {editedRewards.map((reward, index) => (
                <View key={index} style={styles.rewardRow}>
                    <TextInput
                        style={styles.rewardInput}
                        placeholder="Stamps required (i.e: 5)"
                        value={reward.stamps}
                        onChangeText={(text) => {
                            const updatedRewards = [...editedRewards];
                            updatedRewards[index].stamps = text;
                            setEditedRewards(updatedRewards);
                        }}
                    />
                    <TextInput
                        style={styles.rewardInput}
                        placeholder="Reward (i.e: 1 free drink)"
                        value={reward.reward}
                        onChangeText={(text) => {
                            const updatedRewards = [...editedRewards];
                            updatedRewards[index].reward = text;
                            setEditedRewards(updatedRewards);
                        }}
                    />
                    <TouchableOpacity style={styles.removeRewardButton} onPress={() => removeRewardField(index)}>
                        <Ionicons name="ios-trash-outline" size={24} color="red" />
                    </TouchableOpacity>
                </View>
            ))}

            {/* Add Reward button */}
            <TouchableOpacity style={styles.addRewardButton} onPress={addRewardField}>
                <Text style={styles.buttonText}>Add Reward</Text>
            </TouchableOpacity>

            <TouchableOpacity style={styles.createPlanButton} onPress={handleEditPlan}>
                <Text style={styles.buttonText}>Save Changes</Text>
            </TouchableOpacity>
        </KeyboardAvoidingView>
    );
};

export default EditPlanScreenView;

const styles = StyleSheet.create({
    container: {
        flex: 1,
        padding: 20,
    },
    heading: {
        fontSize: 24,
        marginBottom: 20,
        textAlign: 'center',
    },
    tag: {
        fontSize: 15,
        marginBottom: 5,
    },
    input: {
        marginBottom: 15,
        padding: 10,
        borderWidth: 1,
        borderColor: '#ccc',
        borderRadius: 5,
    },
    colorPicker: {
        flexDirection: 'row',
        alignItems: 'center',
    },
    createPlanButton: {
        backgroundColor: '#0782F9',
        padding: 15,
        borderRadius: 10,
        alignItems: 'center',
        marginTop: 10,
    },
    addRewardButton: {
        backgroundColor: '#14A44D',
        padding: 10,
        borderRadius: 5,
        alignItems: 'center',
        marginTop: 10,
        marginBottom: 60,
    },
    buttonText: {
        color: 'white',
        fontWeight: '700',
        fontSize: 16,
    },
    rewardRow: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginBottom: 5,
    },
    rewardInput: {
        flex: 1,
        marginRight: 5,
        padding: 10,
        borderWidth: 1,
        borderColor: '#ccc',
        borderRadius: 5,
    },
    removeRewardButton: {
        alignSelf: 'flex-start',
        padding: 10,
        borderRadius: 5,
    },
});
