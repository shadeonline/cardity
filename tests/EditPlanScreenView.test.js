import React from 'react';
import { render } from 'react-native-testing-library';
import EditPlanScreenView from '../screens/EditPlanScreenView';
import { NavigationContainer } from '@react-navigation/native';

// Custom mock for useNavigation
const mockNavigation = {
    goBack: jest.fn(),
};

// Mock dependencies
jest.mock('@react-navigation/stack', () => ({
    useFocusEffect: jest.fn(),
    useNavigation: () => mockNavigation, // Use the custom mock here
}));

jest.mock('../firebase', () => ({
    auth: {
        currentUser: {
            email: 'admin@example.com', // Mock the user's email here
        },
    },
}));

jest.mock('firebase/auth', () => ({
    signOut: jest.fn(), // Mock the signOut function here
}));

// Mock firebaseFunctions
jest.mock('../firebaseFunctions', () => ({
    firebaseEditLoyaltyPlan: jest.fn(),
}));

describe('EditPlanScreenView Component', () => {
    it('matches snapshot', async () => {
        // Mock data
        const programData = { "description": "Earn free drinks at Stag Bar now!", "id": "n2KSNnpuLIciipIegOzl", "loyaltyCard": { "cardName": "Stag Card", "color": "#792cff", "id": "kyK8L7oqA5SsEHSiZqRH" }, "rewards": { "10": "Free IPA", "5": "Free stout beer" }, "storeName": "Stag bar", "task": "Purchase $10 worth of product at store" }

        // Render the EditPlanScreenView component with the mock data
        const { toJSON } = render(
            <NavigationContainer>
                <EditPlanScreenView route={{ params: { programData } }} />
            </NavigationContainer>
        );

        // Use Jest's expect to match the snapshot
        expect(toJSON()).toMatchSnapshot();
    });
});
