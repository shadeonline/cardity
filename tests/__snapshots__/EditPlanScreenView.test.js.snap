// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPlanScreenView Component matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "flex": 1,
        "padding": 20,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <Text
    style={
      {
        "fontSize": 15,
        "marginBottom": 5,
      }
    }
  >
    Card Name:
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="i.e: Stag Nite Bar Card"
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 5,
        "borderWidth": 1,
        "marginBottom": 15,
        "padding": 10,
      }
    }
    value="Stag Card"
  />
  <Text
    style={
      {
        "fontSize": 15,
        "marginBottom": 5,
      }
    }
  >
    Store Name:
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="i.e: Stag Nite Bar"
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 5,
        "borderWidth": 1,
        "marginBottom": 15,
        "padding": 10,
      }
    }
    value="Stag bar"
  />
  <Text
    style={
      {
        "fontSize": 15,
        "marginBottom": 5,
      }
    }
  >
    Loyalty Plan Description:
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="i.e: Earn great and awesome reward with this plan"
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 5,
        "borderWidth": 1,
        "marginBottom": 15,
        "padding": 10,
      }
    }
    value="Earn free drinks at Stag Bar now!"
  />
  <Text
    style={
      {
        "fontSize": 15,
        "marginBottom": 5,
      }
    }
  >
    Task:
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="i.e: Earn great and awesome reward with this plan"
    style={
      {
        "borderColor": "#ccc",
        "borderRadius": 5,
        "borderWidth": 1,
        "marginBottom": 15,
        "padding": 10,
      }
    }
    value="Purchase $10 worth of product at store"
  />
  <Text
    style={
      {
        "fontSize": 15,
        "marginBottom": 5,
      }
    }
  >
    Color of loyalty card:
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 20,
          "marginTop": 5,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#2541b2",
            "borderColor": "green",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#1768ac",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#06bee1",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffc2c7",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#ffb1f2",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#dd9eff",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#c966ff",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#792cff",
            "borderColor": "transparent",
            "borderRadius": 20,
            "borderWidth": 5,
            "height": 40,
            "opacity": 1,
            "width": 40,
          }
        }
      />
    </View>
  </View>
  <Text
    style={
      {
        "fontSize": 15,
        "marginBottom": 5,
      }
    }
  >
    Rewards:
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 5,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Stamps required (i.e: 5)"
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "marginRight": 5,
          "padding": 10,
        }
      }
      value="5"
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Reward (i.e: 1 free drink)"
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "marginRight": 5,
          "padding": 10,
        }
      }
      value="Free stout beer"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignSelf": "flex-start",
          "borderRadius": 5,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 5,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Stamps required (i.e: 5)"
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "marginRight": 5,
          "padding": 10,
        }
      }
      value="10"
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Reward (i.e: 1 free drink)"
      style={
        {
          "borderColor": "#ccc",
          "borderRadius": 5,
          "borderWidth": 1,
          "flex": 1,
          "marginRight": 5,
          "padding": 10,
        }
      }
      value="Free IPA"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignSelf": "flex-start",
          "borderRadius": 5,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text />
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#14A44D",
        "borderRadius": 5,
        "marginBottom": 60,
        "marginTop": 10,
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "700",
        }
      }
    >
      Add Reward
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#0782F9",
        "borderRadius": 10,
        "marginTop": 10,
        "opacity": 1,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "700",
        }
      }
    >
      Save Changes
    </Text>
  </View>
</View>
`;
